src/EventListener/ExceptionListener.php

```diff
@@ -0,0 +1,307 @@
+        }
+
+        // FIXME: log the exception here if we do not find a better place to log via monolog from outside.
+
+        if (!$this->renderErrorScreens) {
+            return;
+        }
+
+        // Search if any exception in the chain is implementing a http exception.
+        $response = $this->checkHttpExceptions($exception);
+        if ($response instanceof Response) {
+            $event->setResponse($this->setXStatusCode($response));
+
+            return;
+        }
+
+        // If still nothing worked out, we wrap it in a plain "internal error occured" message.
+        $event->setResponse($this->setXStatusCode(new Response($this->renderErrorTemplate('be_error'), 500)));
+    }
+
+    /**
+     * Ensure we have a proper X-Status-Code set on a response.
+     *
+     * Setting the status code is needed as otherwise the kernel will not know we handled the exception and will set
+     * an status code of 500 ignoring the status of our response.
+     *
+     * @param Response $response
+     *
+     * @return Response
+     */
+    private function setXStatusCode(Response $response)
+    {
+        if (!$response->headers->has('X-Status-Code')) {
+            $response->headers->set('X-Status-Code', $response->getStatusCode());
+        }
+
+        return $response;
+    }
+
+    /**
+     * Check if the exception or any exception in the chain implements the response exception interface.
+     *
+     * @param \Exception $exception The exception to walk on.
+     *
+     * @return null|Response
+     */
+    private function checkResponseException($exception)
+    {
+        do {
+            if ($exception instanceof ResponseExceptionInterface) {
+                return $exception->getResponse();
+            }
+        } while (null !== ($exception = $exception->getPrevious()));
+
+        return null;
+    }
+
+    /**
+     * Check if the exception or any exception in the chain inherits from HttpException and is understood by us.
+     *
+     * @param \Exception $exception The exception to walk on.
+     *
+     * @return null|Response
+     */
+    private function checkHttpExceptions(\Exception $exception)
+    {
+        do {
+            if ($response = $this->checkHttpException($exception)) {
+                return $response;
+            }
+        } while (null !== ($exception = $exception->getPrevious()));
+
+        return null;
+    }
+
+    /**
+     * Try to render a Contao 404 page first.
+     *
+     * @param \Exception $exception The exception to check.
+     *
+     * @return null|Response
+     *
+     * @throws \Exception When anything went really haywire during rendering of the page.
+     */
+    private function tryToRenderContao404($exception)
+    {
+        // FIXME: I guess it is hopeless to render a 404 page from within a root page when no root page is published.
+        // We might want to skip RootNotFoundHttpException then and also NoPagesFoundHttpException is an candidate.
```

Comment:
But there are places (even in Contao 3) where it is impossible to render the 404 page, because it would end in an endless loop.